proposals/lip_define_state_and_state_transitions_of_dpos_module.md

```diff
@@ -0,0 +1,1163 @@
+
+    # Reorder the validator list
+    shuffledValidatorAddresses = sort validatorsAddresses where address1 < address2 if (roundHash(address1) < roundHash(address2))
+                                 or ((roundHash(address1) == roundHash(address2)) and address1 < address2)         
+    
+    return shuffledValidatorAddresses
+```
+
+
+### Block Processing
+
+
+#### After Genesis Block Execution
+
+After the genesis block `b` is executed, the following logic is executed:
+
+```python
+create the entry in the genesis data substore with 
+store value = {
+    "height": b.header.height, 
+    "initRounds": b.header.assets.initRounds, 
+    "initDelegates": b.header.assets.initDelegates
+} serialized using genesisDataStoreSchema
+
+# set the inititial delegates in the BFT module
+bftWeights = [
+    {"address": address, "bftWeight": 1} 
+    for address in initDelegates
+    sorted by lexicographically by address
+]
+
+# compute the initial BFT threshold 
+initBFTThreshold = floor(2/3 * length(initDelegates)) + 1
+
+BFT.setBFTParameters(initBFTThreshold,
+                     initBFTThreshold,
+                     bftWeights)
+                     
+# set the inititial delegates in the validators module                 
+validators.setGeneratorList(initDelegates)
+```
+
+
+#### After Block Execution
+
+After a block `newBlock` is executed, the properties related to missed blocks are updated according to [Delegate Productivity][LIP-0023-delegateProductivity]. This logic is recapitulated below:
+
+```python
+newHeight = newBlock.header.height
+# previousTimestamp is the value in the previous timestamp substore
+missedBlocks = validators.getGeneratorsBetweenTimestamps(previousTimestamp, newBlock.header.timestamp)
+
+# remove the start and end blocks, as those are not missed
+missedBlocks[validators.getGeneratorAtTimestamp(previousTimestamp)] -= 1
+missedBlocks[validators.getGeneratorAtTimestamp(newBlock.header.timestamp)] -= 1
+
+for address in missedBlocks:
+    delegateStore(address).consecutiveMissedBlocks += missedBlocks[address]
+
+    # the rule below was introduced in LIP 0023
+    if (delegateStore(address).consecutiveMissedBlocks > FAIL_SAFE_MISSED_BLOCKS
+        and newHeight - delegateStore(address).lastGeneratedHeight > FAIL_SAFE_INACTIVE_WINDOW):
+        delegateStore(address).isBanned = True
+
+delegateStore(newBlock.header.generatorAddress).consecutiveMissedBlocks = 0
+delegateStore(newBlock.header.generatorAddress).lastGeneratedHeight = newHeight
+
+previousTimestamp = newBlock.header.timestamp
+```
+
+After an end-of-round block `b` is executed (`isEndOfRound(b.header.height) == True`), the following logic is executed (this must be done after the properties related to missed blocks are updated):
+
+```python
+roundNumber = roundNumber(b.header.height)
+currentWeights = {}
+for address being a storeKey in delegate substore and isBanned(address) == False:
+    currentWeights[address] = delegateWeight(address, b.header.height)} 
```

Comment:
`delegateWeight(address, b.header.height)}`, what is this lonely curly bracket is doing there 🙂 ?

Reply:
chilling?
good catch, I removed it 🙂.